include/nil/blueprint/components/systems/snark/plonk/placeholder/fri_cosets.hpp

```diff
@@ -332,107 +367,142 @@ namespace nil {
-                cs3 = {six_block[0]};
-                if (sixb_rows > 1) { // if the first 6-block is a regular middle 6-block, otherwise there's no "previous"
-                    cs3.push_back(six_block[1]);
-                    cs3.push_back(six_block[2]);
+                std::size_t last = (last_l > 0)? last_l : l; // The number of the last 9-block in the row
+                cs3 = {nine_block[0],nine_block[1],nine_block[2]};
+                if (nineb_rows > 1) { // if the first 9-block is a regular middle 9-block, otherwise there's no "previous"
+                    cs3.push_back(nine_block[3]);
+                    cs3.push_back(nine_block[4]);
                 } else {
-                    cs3.push_back(first_W2);
+                    cs3.push_back(first_add_W1);
+                    cs3.push_back(first_add_W2);
+                    cs3.push_back(first_W4);
                 }
-                cs3.insert(cs3.end(),std::next(six_block.begin(),3),std::next(six_block.begin(),6*(last-1)+3));
-                cs3.push_back(var(component.W(6*(last - 1) + 3),0) - var(component.W(6*(last - 1) + 2),0));
-                cs3.push_back(six_block[6*(last-1)+4]);
-                cs3.push_back(six_block[6*(last-1)+5]);
-
-                if (last_l > 0) { // there are octets on the transition line
-                    constraint_type mid = var(component.W(0),1);
-
-                    for(std::size_t j = WA-1; j > 6*last_l; j--) {
-                        mid *= 8;
+                cs3.insert(cs3.end(),std::next(nine_block.begin(),5),std::next(nine_block.begin(),9*(last-1)+5));
+                cs3.push_back(var(component.W(9*(last - 1) + 5),0) - var(component.W(9*(last - 1) + 4),0));
+                cs3.push_back(nine_block[9*(last-1)+6]);
+                cs3.push_back(nine_block[9*(last-1)+7]);
+                cs3.push_back(nine_block[9*(last-1)+8]);
+
+                if (last_l > 0) { // there are bits on the transition line
+                    constraint_type mid  = var(component.W(0),1),
+                                    mid1 = var(component.W(1),1);
+
+                    for(std::size_t j = WA-3; j > 9*last_l; j -= 3) {
+                        mid *= 2;
                         mid += var(component.W(j),0);
+                        mid1 *= 2;
+                        mid1 += var(component.W(j+1),0);
                     }
-                    mid -= var(component.W(6*last_l),0);
+                    mid  -= var(component.W(9*last_l),0);
+                    mid1 -= var(component.W(9*last_l + 1),0);
                     cs3.push_back(mid);
+                    cs3.push_back(mid1);
+
+                    // the sign bit shoudl be just a copy from the next block
+                    var Wl2     = var(component.W(9*last_l + 2),0),
+                        Wl2next = var(component.W(9*last_l + 5),0);
+                    cs3.push_back( Wl2 - Wl2next );
```

Comment:
Given that we have to assert this equality in only a single row a copy constraint should be more efficient? Unsure.

Reply:
It seems to me, that moving this to copy constraints slightly disrupts the logic. We would then have a 3-block with some parts encoded as polynomial constraints and other parts as copy constraits, whereas other 3-blocks are all encoded with just poly-constraints. I would move this to copy constraints only if there is a clear performance advantage.